### PR TITLE
json: Added element descriptor macro and C++ support progress

### PR DESCRIPTION
**Pull request by Sir-Branch**
Opening this PR to start the discussion, code formatting is not perfect.

 I'm working towards cleaning up the JSON library and making it more maintainable, on my high priorities is compatibility with C++.

What are the Zephyr guidelines towards having empty macro arguments? In this specific use case the union_ parameter in JSON_ELEMENT_DESCR may be empty, this is to insure a single point of use for ``` (struct json_obj_descr[])```

The underlying issue which motivates this PR is issue #18551, in this case macros that utilize ``` struct json_obj_descr[] ``` aren't able to run on C++. A possible patch is to replace all these occurrents for ``` const struct json_obj_descr[1] ```.

### CODE REVIEW

**Review by martinjaeger — COMMENTED** [review]
If the new macros should not be used from the application, they should probably be prefixed with `Z_` so that they are excluded from the documentation (see fb4d9c04111b145e08fd6848d7a4d595705afafe).

Otherwise looks good to me.

Empty arguments seem to be valid C99, [see here](https://stackoverflow.com/questions/7666344/are-empty-macro-arguments-legal-in-c11). So I don't see any issue with that.

**Comment by Sir-Branch** [comment]
I appreciate the insight, thanks for the help!

- I went ahead and prefixed with Z_
- Improved C++ support by changing from ` struct json_obj_descr[]` to `const struct json_obj_descr[1]`
- Rebased to pass CI

File: include/data/json.h
Comment (martinjaeger) [inline include/data/json.h:186]: Whether `const struct json_obj_descr[]` or `const struct json_obj_descr[1]` is better/valid is beyond my C vs. C++ knowledge. Maybe @pabigot could have a quick look, as you provided related PRs for Bluetooth?

Comment (pabigot) [reply include/data/json.h:186]: Since when I fixed the C++ problem with using an address of temporary I used `type[]` I don't know of any reason why it should change to `type[1]`, but newer versions of GNUC might have a problem.  In context `[]` should just cause the compiler to infer the array length from the single-element value.

Comment (Sir-Branch) [reply include/data/json.h:186]: Agreed, there really shouldn't be a difference between ```type[]``` and ```type[1]``` as compiler will infer the length. Specifying the length could cause maintenance issue down the road, I'll update to ```const struct json_obj_descr[]```. I really  appreciate the insight

**Review by martinjaeger — APPROVED** [review]

**Comment by Sir-Branch** [comment]
@martinjaeger rebased to avoid conflict due to #38061: Updated use of `JSON_TOK_LIST_` to `JSON_TOK_ARRAY_`

File: include/data/json.h
Comment (cfriedt) [inline include/data/json.h:182]: May be good to apply the `@internal` tag as well

File: include/data/json.h
Comment (cfriedt) [inline include/data/json.h:201]: May be good to apply the `@internal` tag as well

File: include/data/json.h
Comment (cfriedt) [inline include/data/json.h:215]: May be good to apply the `@internal` tag as well

**Comment by Sir-Branch** [comment]
@cfriedt Thank you for the review have updated accordingly.

I really appreciate all the help you and @martinjaeger have provided, you've both made the process of contributing to Zephyr seamless.

**Comment by Sir-Branch** [comment]
@mrfuchs - Would it be possible to get a review?

File: include/data/json.h
Comment (mrfuchs) [inline include/data/json.h:523]:
```suggestion
					Z_JSON_DESCR_OBJ(elem_descr_, elem_descr_len_)), \
```

**Review by mrfuchs — CHANGES_REQUESTED** [review]
`JSON_OBJ_DESCR_OBJ_ARRAY_NAMED` needs to be fixed, otherwise LGTM.

File: include/data/json.h
Comment (mrfuchs) [inline include/data/json.h:521]:
```suggestion
				.element_descr = Z_JSON_ELEMENT_DESCR(struct_, len_field_, \
```

Comment (Sir-Branch) [reply include/data/json.h:521]: Nice catch, thank you!

File: include/data/json.h
Comment (Sir-Branch) [reply include/data/json.h:523]: Thank you, this one totally slipped!

**Review by mrfuchs — APPROVED** [review]

**Comment by Sir-Branch** [comment]
@mrfuchs thank you for the review. Have updated accordingly 😄

**Comment by Sir-Branch** [comment]
@cfriedt do you think this can be merged in? Not sure why @mrfuchs approval isn't counting, even though he's a collaborator. Maybe he doesn't have write access? (Apologies if this is obvious, this is new to me)

**Comment by cfriedt** [comment]
@mbolivar-nordic - I think that @mrfuchs is a collaborator for JSON. Any idea why have a approval might not work?

**Review by cfriedt — APPROVED** [review]

**Pull request merged**
